Refactored parsing; Added reducing of pricebooks and inventories; added support of passing multiple catalogs using patterns, added types definition and async code-style

hryhola opened the pull request:
Added:
- reducing of pricebooks and inventories
- support of passing multiple catalogs using patterns (using 'globby' library)
- added types definition (in types.ts file) and async-await code-style as a wrapper for `on('end', callback).start()`
- jsdoc comments for class methods pointing overrides

Refactored:
- switched from CommonJS to ES modules
- renamed class names for more intuitiveness
- registry cache writing and reading
- divided scripts to separate files

Removed:
- unused, duplicated functions

Commit:
Implemented 'variation-group' product type handling

Commit:
Reverted 'variation-group' product type

Commit:
Refactored catalog worker inheritance, jsdoc comments added, code style fixed

Commit:
Import fixed, type fixes

Commit:
Path variable renames

Commit:
Added alias variable for temp master file

Commit:
Moved code block to proper place

Commit:
Renamed MasterCatalogFilter to XMLFilterWriter

Commit:
Renamed NavigationCatalogRegistry to NavigationAssignmentsRegistry, added minifiedNavigationPath

Commit:
Renamed CatalogRegistry to ProductsRegistry

Commit:
Created Registry class

Commit:
Implemented filtering of category definitions in navigation catalog

Commit:
Added async-styled execution possibility

Commit:
Added filter option in XMLFilterWriter constructor

Commit:
Added types for CatalogReducerConfig

Commit:
Fixed paths

Commit:
Added generation of inventory catalog

Commit:
Added out postfix option

Commit:
Added generation of minified price books

Commit:
Divided reducers into separate files, discarded usage of 'out file src' option

Commit:
Async reducers fix

Commit:
Added 'update' behaviour

Commit:
ES module usage added

Commit:
Added patterns usage for inventories and pricebooks

Commit:
Fixed types import

Commit:
Added support of multiple navigation catalogs

Commit:
Removed parsing of the same file patterns multiple times

Commit:
Updated test data, fixed cache rendering

Commit:
Added support of multiple master and navigation catalogs for test data

Commit:
Added inventory test data

Commit:
Added price books test data

Commit:
Changed prefix for second catalog product to 'x-'

Commit:
Fixed errors log

Commit:
Added node_modules and package-lock.json to gitignore

Commit:
Updated readme

Commit:
productFilter type fixed

rzhornyk commented on .vscode/launch.json:
```diff
@@ -0,0 +1,18 @@
+{
```
This should be in gitignore and should not be committed

rzhornyk commented on README.md:
```diff
@@ -18,31 +30,35 @@ Please update your project `package.json` with configuration:
 {
     "catalogReducer": {
```
Can we move it outside of package.json?
Actually for backward compatibility supports both options but preferable is to use separate file

rzhornyk commented on README.md:
```diff
@@ -5,10 +5,22 @@
+
+## See how it works
+
+Execute `npm run test-reduce` and see the result of processing of test data in sfcc-catalog-reducer/test-data folder.
```
Seems something went wrong with executing this command, it shows "undefined" in catalog names http://prntscr.com/T06Z5sYPHxZY

rzhornyk commented on lib/tools/cleanup.js:
```diff
@@ -1,22 +1,65 @@
-}
\ No newline at end of file
+/**
+ * @param {Array<string>} files
```
Add description what this method do.

rzhornyk commented on lib/tools/cleanup.js:
```diff
@@ -1,22 +1,65 @@
+        dirs.forEach(folder => {
+            log(`Cleanup folder ${folder}`);
+
+            removeFilesExcept(folder, files, (file) => !isOutPostfixFile(file));
```
Not fully understand logic behind this operation. Could you explain?

hryhola replied:
Do not delete reduced files that we just created.
For example, if we set behavior to 'createNew' and the ending part is '_reduced', we don't want files that end with '_reduced' to be removed on clean up at the end of script execution.

rzhornyk commented on lib/catalog/registry/ProductsRegistry.js:
```diff
@@ -0,0 +1,37 @@
+     * @returns {boolean}
+     */
+    hasProduct (id) {
+        return id in this.products;
```
This method is no longer used and has not actual logic because you now have this.cache.products

rzhornyk commented on lib/catalog/registry/NavigationCategoriesRegistry.js:
```diff
@@ -0,0 +1,78 @@
+     * @param {string} categoryId
+     * @param {string | null} productId
+     */
+    addCategory (categoryId, parrentId) {
```
parent with one "r"

rzhornyk commented on lib/tools/reducers/master.js:
```diff
@@ -0,0 +1,23 @@
+
+    const masterFilterByAssignments = new XMLFilterWriter(tempFileWithFilteredProducts, outFile);
+
+    return masterFilterByAssignments.startAsync('category-assignment', productFilter);
```
it not return any value from promise, so here better to write "await" instead of "return"

rzhornyk commented on lib/tools/reducers/navigation.js:
```diff
@@ -0,0 +1,33 @@
+
+            const navigationFilterByCategories = new XMLFilterWriter(tempFileWithFilteredProducts, outFile);
+
+            return navigationFilterByCategories.startAsync('category', tag => {
```
"await" instead of "return"

rzhornyk commented on package.json:
```diff
@@ -1,13 +1,21 @@
 {
     "name": "sfcc-catalog-reducer",
-    "version": "1.1.1",
+    "version": "2.0",
```
2.0.0

Commit:
Removed .vscode folder

Commit:
Added .vscode folder to gitignore

Commit:
Fixed setting of cache name

Commit:
Set version to 2.0.0

rzhornyk commented on configs/default.json:
```diff
@@ -14,4 +11,4 @@
             "standard": 0
```
What about variation groups?
How can we ensure that variation groups will be not removed from catalog?

hryhola replied:
Variation groups will not be removed if the corresponding master product will not be removed too, works as expected. I'll add product with variation groups to test data.

Commit:
Removed redundant return of void promise

Commit:
Misspell fixed

Commit:
Added support of configuration in catalogReducerConfing.json file

Commit:
Added master and navigation catalogs amount check

Commit:
Added description for cleaner

Commit:
Removed unused method

Commit:
Added master prodct with variation groups to test data

Commit:
Removed skipping of unused categories in navigation catalog

Commit:
Type added

Commit:
Added $ prefix for default category option

Commit:
Moved variation group product to deals category

Commit:
Refactored parsing, discarded cache usage, added subpath imports

Commit:
Added masterWithVariationGroups type, added Set usage for savedProductsDependencies list

Commit:
Fixed parsing of standard products

Commit:
Added gray color for time log, added beep sound at script end

Commit:
Deleted test-cache-reset script

Commit:
Added termination of read stream if a container for saved products is full

Commit:
Added name for xml tools for logging

Commit:
Replaced object usage with Set usage as a continer for reduced products

Commit:
Found bug with standard products

Commit:
Fixed parsing of standard products, added check for online-flag

Commit:
Added ProductsContainer class and XMLProductDefinition class

Commit:
Added parsing on multiple categories at the same time

Commit:
Added saving products for default category

Commit:
Logger renames

Commit:
Moved reducers to lib folder

Commit:
Added writing for default category

Commit:
Added parsing of multiple tags, fixed reducing for default category

Commit:
Added parsing of multiple tags, fixed reducing for default category

Commit:
Merge branch 'feature/parsing-refactoring' of https://github.com/hryhola/sfcc-catalog-reducer into feature/parsing-refactoring

Commit:
Removed temp file creating for master reducer

Commit:
Added parsing of master products even if the have no assignment in navigation catalog

Commit:
Added 'keepAsItIs' reducing option for a category

Commit:
Added DependancyParent class usage instead of full XMLProductDefinition for storing data about product in ram

Commit:
Added predefined products option for catalog reducer config

Commit:
Added generation of inventory allocation for missing records

Commit:
Removed inventory record generation for variation-group product type

Commit:
Added generation of missing price records

Commit:
Removed unused console.logs

Commit:
Fixed multiple DependancyParent object creation for the same product

Commit:
Grammar mistakes fixes

Commit:
Grammar mistake rename

Commit:
Eslint fixes

Commit:
File rename (spellcheck fix)

Commit:
Updated readme

hryhola committed:
Merge pull request #1 from hryhola/feature/parsing-refactoring

Refactored parsing functionality

hryhola commented:
Added:
- Generating of missing records for a product:
  - allocation in inventory catalogs
  - price in pricebook catalogs
- Catalog reducer config options:
   - "masterWithVariationGroups" product type
   - "keepAsItIs" option for a category
   - an array of product IDs as option for some product type for some category
   - online-flag check option
- Subpath import usage
- Eslint 

Refactored:
- Parsing functionality: on the fly product types identification, dependency gathering and filtering for the final reduced products list. Previously, such operations were performed on different iterations of parsing.
- Default category ($default) functionality: instead of pointing to total amount of product that should be left (no matter what category, no additional assignment generation), now default category points how much products left in **all** unspecified categories, with generation of corresponding records. Previously, products in the default category could be assigned only to one category, but now after reducing all default category product will be assigned to all unspecific categories.
- Products config: now product config could specify only online-flag check option

Removed:
- Writing to temporary files, caching
- Unused Registry and Worker classes

Commit:
Removed unused types

Commit:
Merge branch 'develop' of https://github.com/hryhola/sfcc-catalog-reducer into develop

Commit:
Removed unused types

Commit:
Merge branch 'feature/parsing-refactoring' of https://github.com/hryhola/sfcc-catalog-reducer into feature/parsing-refactoring

Commit:
Merge branch 'feature/parsing-refactoring' of https://github.com/hryhola/sfcc-catalog-reducer into develop

Commit:
Rename spell fix

Commit:
Spell fix

rzhornyk commented on configs/test.json:
```diff
@@ -0,0 +1,46 @@
+                "product-set-3"
+            ]
+        },
+        "category-1": "keepAsItIs"
```
"keepAsIs" sounds better

rzhornyk commented on lib/model/CategoriesProductsContainer.js:
```diff
@@ -0,0 +1,108 @@
+import { inspect } from 'util';
+
+import ProductsConainter from './ProductsContainer.js';
```
Typo

rzhornyk commented on README.md:
```diff
@@ -5,44 +5,71 @@
+            "set": 5,
+            "bundle": 5,
+            "standard": 10
         },
```
Here not described "keepAsIs"

rzhornyk commented on lib/model/CategoriesProductsContainer.js:
```diff
@@ -0,0 +1,108 @@
+         */
+        this.categories = {};
+
+        const allObjectConfigs = filterKeys(categoryConfigs, (key) => typeof categoryConfigs[key] !== 'string');
```
Category config can be only in one case as string, when we set "keepAsIs" to it.
I spend 5 minutes to understand why this filtering needs.
Could you add inline comment with explanation or rewrite code to explicitly filter "keepAsIs"?

Commit:
Removed duplicated field in config example

rzhornyk commented on lib/model/ProductsContainer.js:
```diff
@@ -0,0 +1,152 @@
+    }
+
+    isFull() {
+        return Object
```
To cache this value, and invalidate on executing "add" method.

rzhornyk commented on lib/model/CategoriesProductsContainer.js:
```diff
@@ -0,0 +1,108 @@
+        return this.categories[categoryID].isFull();
+    }
+
+    isEveryCategoryFull() {
```
You execute this a lot of time, and this is heavy operation to recalculate it every time.
Suggest to cache this value and invalidate on add method
```
isEveryCategoryFull() {
    if (typeof this.cachedEveryCategoryFullFlag) {
        return this.cachedEveryCategoryFullFlag;
    }

    this.cachedEveryCategoryFullFlag = // current logic

    return this.cachedEveryCategoryFullFlag;
}
```
and in "add" method
``` 
add(categoryID, product) {
     this.categories[categoryID].add(product);

     this.cachedEveryCategoryFullFlag = undefined;

     return this;
}
````

rzhornyk commented on .gitignore:
```diff
@@ -1,3 +1,25 @@
+node_modules
+package-lock.json
```
Why we ignore package-lock.json?

rzhornyk commented on lib/workers/ProductDefinitionWorker.js:
```diff
@@ -0,0 +1,305 @@
+                                    return;
+                                }
+
+                                const patentCategories = whitelistedProductCategories;
```
typo

rzhornyk commented on lib/workers/ProductDefinitionWorker.js:
```diff
@@ -0,0 +1,305 @@
+                const product = new XMLProductDefinition(raw, attributes, skipList);
+
+                const onlineCheckResult = onlineFlagCheck
+                    ? product.onlineFlag === 'true'
```
product.onlineFlag not take to account if product was online only for some site
For example 
```
<online-flag>false</online-flag>
<online-flag site-id="site-id">true</online-flag>
```

rzhornyk commented on lib/xml/XMLProductDefinition.js:
```diff
@@ -0,0 +1,142 @@
+const DEPENDENCY_REGEX = /product-id="(.+?)"/g;
+const VARIATION_GROUP_REGEX = /variation-group\sproduct-id="(.+?)"/g;
+const ONLINE_FLAG_REGEX = /<online-flag>([a-z]+)<\/online-flag>/;
```
to update to handle at least online for one site
/<online-flag.*?>true<\/online-flag>/

rzhornyk commented on lib/reducers/navigation.js:
```diff
@@ -0,0 +1,62 @@
+            });
+        }
+
+        await navigationFilterWriter.startAsync('category-assignment', /** @type {import('#types').XMLTagFilter} */(tag) => {
```
There also exists tag
```
<recommendation source-id="test-shelby" source-type="product" target-id="101367434" type="1"/>
```
Filter it as well.

rzhornyk commented on lib/reducers/priceBook.js:
```diff
@@ -0,0 +1,52 @@
+        /**
+         * If it's the last file
+         */
+        if (index === files.length - 1) {
```
you should write prices not only into last file.
Each different price book file belong to different currency, or different price book type (list or sale)
Price should be present in each list pricebook.
You can detect if pricebook is list by checking pricebooks>pricebook>header>parent tag. If it is not present or empty then it is list price.

Commit:
Merge branch 'develop' of https://github.com/hryhola/sfcc-catalog-reducer into develop

Commit:
Renamed 'keepAsItIs' to 'keepAsIs'

Commit:
Typo fixed

Commit:
Added description of 'keepAsIs' option

Commit:
Added comment describing gathering object configs

Commit:
Added caching for ProductsContainer.isFull method

Commit:
Renamed 'result' to 'flag'

Commit:
Added caching for CategorizedProductsContainer.isEveryCategoryFull

Commit:
Removed package-lock.json from gitignore

Commit:
Typo fixed

Commit:
Added online-flag check in case of multiple sites configuration

Commit:
Fixed online-flag handling

Commit:
Added filtering of 'recommendation' tag in navigation catalogs

Commit:
Added generating of all price-tables in list pricebooks, removed generating of price-tables for non-list pricebooks

Commit:
Added 'recommendation' tags to test master catalogs

Commit:
Added filtering of 'recommendation' tags for master catalogs

Commit:
Fixed eslint issues, created skip words list

Commit:
Added linting for all folders and ts files

Commit:
Updated algorithm description in README.md

Commit:
Renamed CategoriesProductsContainer.js to CategorizedProductsContainer.js

Commit:
Added a dot

rzhornyk commented on README.md:
```diff
@@ -4,45 +4,84 @@
 
 ## How it works?
```
To add new section 
```
## Idea

Usually production catalog is huge and SFCC sandboxes cannot handle a big amount of catalog data without performance degradation. On the other hand, a sandbox is an instance for developing a site and it does not require an entire catalog to be present on it. 
Will be enough to have few properly configured categories and all types of products. All other categories in the navigation catalog can have assigned some dummy products, just for the proper rendering navigation menu. Based on such assumption was created this tool, that allows specifying a few categories that will be kept untouched (see "keepAsIs" config) or preserve in it specific products(by specifying IDs) that are for sure properly configured, or just set amount of product per each product type (master, standard, set, bundle) to be preserved in each category.
```

hryhola replied:
Added

rzhornyk commented on configs/test.json:
```diff
@@ -0,0 +1,48 @@
+            "./testdata/navigation/navigation2.xml"
+        ],
+        "inventories": [
+            "testdata/inventory/inventory1.xml"
```
In code you have committed inventory2.xml file as well.
Should it be present here?

hryhola replied:
The idea was to generate missing records in inventory1.xml, but I guess it would make more sense just to remove data from inventory2 and generate it there.

Commit:
Added empty inventory2.xml as placeholder for generating missing inventory records

Commit:
Removed unused options from catalogReducerConfig example

Commit:
Added Idea section in README.md

Commit:
Adde src check

Commit:
Set default values

Commit:
Default values fix

Commit:
category warn added

Commit:
Added wrong categoriesConfig exit cases

Commit:
Removed product assignment generation for non-master products
